### ISSUE TITLE
The ApplicationAccessibilityEnabled hack may turn out to be a very bad idea

### ISSUE THREAD

**Issue by 0xced** [domesticcatsoftware/DCIntrospect#38]
It seems that the hack enabling accessibility without the accessibility inspector in the simulator has very bad unintended side effects.

Setting the `ApplicationAccessibilityEnabled` preference to `true` in `~/Library/Application Support/iPhone Simulator/5.0/Library/Preferences/com.apple.Accessibility.plist` will change the behavior of `UITableView` instances for **all apps** in the simulator in a very unexpected way.

The data source method `- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath;` and the delegate method `- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath;` will be called for **all** index paths, even for cells which are not visible when the app is totally idle! Deleting the `ApplicationAccessibilityEnabled` preference restores the expected behavior of `UITableView`.

Here are the relevant backtraces.

```
#0  -[TableViewController tableView:cellForRowAtIndexPath:] ()
#1  0x00444e0f in -[UITableView(UITableViewInternal) _createPreparedCellForGlobalRow:withIndexPath:] ()
#2  0x004455d6 in -[UITableView(UITableViewInternal) _createPreparedCellForRowAtIndexPath:] ()
#3  0x0c96c0ad in -[UITableViewAccessibility(Accessibility) accessibilityCellForRowAtIndexPath:] ()
#4  0x0c97f60b in -[UITableViewCellAccessibilityElement tableViewCell] ()
#5  0x0c98012a in -[UITableViewCellAccessibilityElement isAccessibilityElement] ()
#6  0x0b774e24 in _appendChildrenToArrayStartingAtIndex ()
#7  0x0b774d3d in _addAXElementsToArrayFromObject ()
#8  0x0b774851 in _appendVendedAXElementsFromUIElements ()
#9  0x0b774774 in _createAXUIElementsFromUIElements ()
#10 0x0b773258 in _copyParameterizedAttributeValueCallback ()
#11 0x0b7aeea6 in _AXXMIGCopyParameterizedAttributeValue ()
#12 0x0b7a9853 in _XCopyParameterizedAttributeValue ()
#13 0x0b7b1e66 in mshMIGPerform ()
#14 0x014e61c5 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ ()
#15 0x0144b022 in __CFRunLoopDoSource1 ()
#16 0x0144990a in __CFRunLoopRun ()
#17 0x01448db4 in CFRunLoopRunSpecific ()
#18 0x01448ccb in CFRunLoopRunInMode ()
#19 0x01c07879 in GSEventRunModal ()
#20 0x01c0793e in GSEventRun ()
#21 0x003aba9b in UIApplicationMain ()
```

```
#0  -[TableViewController tableView:willDisplayCell:forRowAtIndexPath:] ()
#1  0x004454fe in -[UITableView(UITableViewInternal) _createPreparedCellForGlobalRow:withIndexPath:] ()
#2  0x004455d6 in -[UITableView(UITableViewInternal) _createPreparedCellForRowAtIndexPath:] ()
#3  0x0c9db0ad in -[UITableViewAccessibility(Accessibility) accessibilityCellForRowAtIndexPath:] ()
#4  0x0c9ee60b in -[UITableViewCellAccessibilityElement tableViewCell] ()
#5  0x0c9ef12a in -[UITableViewCellAccessibilityElement isAccessibilityElement] ()
#6  0x0ca88e24 in _appendChildrenToArrayStartingAtIndex ()
#7  0x0ca88d3d in _addAXElementsToArrayFromObject ()
#8  0x0ca88851 in _appendVendedAXElementsFromUIElements ()
#9  0x0ca88774 in _createAXUIElementsFromUIElements ()
#10 0x0ca87258 in _copyParameterizedAttributeValueCallback ()
#11 0x0cac2ea6 in _AXXMIGCopyParameterizedAttributeValue ()
#12 0x0cabd853 in _XCopyParameterizedAttributeValue ()
#13 0x0cac5e66 in mshMIGPerform ()
#14 0x014e61c5 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ ()
#15 0x0144b022 in __CFRunLoopDoSource1 ()
#16 0x0144990a in __CFRunLoopRun ()
#17 0x01448db4 in CFRunLoopRunSpecific ()
#18 0x01448ccb in CFRunLoopRunInMode ()
#19 0x01c07879 in GSEventRunModal ()
#20 0x01c0793e in GSEventRun ()
#21 0x003aba9b in UIApplicationMain ()
```

Can anybody else confirm this?

**Comment by sgleadow** [comment]
Sorry it's taken me so long to take a look at this.

I did a test with a basic table view that had about 30 rows, with only 8-9 visible. I tested on device and in the simulator with and without the accessibility hack. In both cases, tableview:cellForRowAtIndexPath: was called 10 times after [tableView reloadData]. That seems like about the right amount for how many are visible. I didn't notice any change with or without the hack.

I wouldn't be surprised if the behaviour you're seeing _is_ a side effect of accessibility, as I guess it needs to indicate what's on and off the screen... but I can't reproduce.

On my current project, we run the accessibility hack on designated test devices plugged into the Jenkins CI box, so I'm not too worried if it does affect scroll performance since it's turning on accessibility is not part of our release build at all. It's still got me a little concerned though!

**Comment by 0xced** [comment]
Closing this. I guess everybody is using Reveal or Xcode's builtin hierarchy viewer nowadays.
